run/opencl/pdf_kernel.cl

```diff
@@ -0,0 +1,695 @@
+#endif
+#endif
+#endif
+
+#if !IS_STATIC_GPU_MASK
+#define GPU_LOC_0 loc0
+#define GPU_LOC_1 loc1
+#define GPU_LOC_2 loc2
+#define GPU_LOC_3 loc3
+#else
+#define GPU_LOC_0 LOC_0
+#define GPU_LOC_1 LOC_1
+#define GPU_LOC_2 LOC_2
+#define GPU_LOC_3 LOC_3
+#endif
+
+	/* Prepare password */
+	prepare234(pwbuf, index, password);
+
+	for (uint mi = 0; mi < NUM_INT_KEYS; mi++) {
+
+		/* Apply GPU-side mask */
+#if NUM_INT_KEYS > 1
+		password[GPU_LOC_0] = int_keys[mi] & 0xff;
+#if MASK_FMT_INT_PLHDR > 1
+#if LOC_1 >= 0
+		password[GPU_LOC_1] = (int_keys[mi] & 0xff00) >> 8;
+#endif
+#endif
+#if MASK_FMT_INT_PLHDR > 2
+#if LOC_2 >= 0
+		password[GPU_LOC_2] = (int_keys[mi] & 0xff0000) >> 16;
+#endif
+#endif
+#if MASK_FMT_INT_PLHDR > 3
+#if LOC_3 >= 0
+		password[GPU_LOC_3] = (int_keys[mi] & 0xff000000) >> 24;
+#endif
+#endif
+#endif
+
+		uint gidx = gid * NUM_INT_KEYS + mi;
+		uint output[32 / 4];
+		uint key[MAX_KEY_SIZE / 8 / 4];
+		uint digest[16 / 4];
+		uint W[(32 + sizeof(pdf_salt->id) + 63 + 9) / 64 * 16]; // Max. 160 bytes, three limbs
+
+		md5_init(key);
+
+		/* Password already padded in prepare() */
+		memcpy_macro(W, password, 32/4);
+		memcpy_macro(W + 32/4, (__constant uint*)pdf_salt->o, 32/4);
+
+		/* The above is always one M-D block */
+		md5_block(uint, W, key);
+
+		W[0] = (uint)pdf_salt->P;
+		uint md5_len = 4;
+
+		memcpy_macro(W + 1, (__constant uint*)pdf_salt->id, pdf_salt->id_len / 4);
+		md5_len += pdf_salt->id_len;
+
+		if (pdf_salt->R >= 4 && !pdf_salt->encrypt_metadata) {
+			W[md5_len / 4] = 0xffffffff;
+			md5_len += 4;
+		}
+
+		W[md5_len / 4] = 0x00000080;
+		for (uint i = md5_len / 4 + 1; i < 14; i++)
+			W[i] = 0;
+		W[14] = (64 + md5_len) << 3;
+		W[15] = 0;
+
+		md5_block(uint, W, key);
+
+		if (pdf_salt->key_length == 40)
+			md5x50_40(key);
+		else
+#if RC4_ANY_KEY_LENGTH
+		if (pdf_salt->key_length == 128)
+#endif
+			md5x50_128(key);
+#if RC4_ANY_KEY_LENGTH
+		else {
+			uint n = pdf_salt->key_length / 8;
+
+			for (uint i = 0; i < 50; i++) {
+				MD5_CTX md5;
+
+				MD5_Init(&md5);
+				MD5_Update(&md5, (uchar*)key, n);
+				MD5_Final((uchar*)key, &md5);
+			}
+		}
+#endif
+
+		uint md5len = 32 + pdf_salt->id_len;
+
+		memcpy_macro(W, padding, 8);
+		memcpy_macro(W + 8, (__constant uint*)pdf_salt->id, pdf_salt->id_len / 4);
+		LASTCHAR(W, md5len, 0x80);
+
+		/* Clean the last M-D block */
+		uint start_clean = md5len / 4 + 1;
+		uint md_len_pos = (start_clean & ~0xfU) + 14 + (((md5len & 63) > 55) ? 16 : 0);
+		for (uint i = start_clean; i < md_len_pos; i++)
+			W[i] = 0;
+		W[md_len_pos] = md5len << 3;
+		W[md_len_pos + 1] = 0;
+
+		md5_init(digest);
+
+		uint *WP = W;
+		for (uint left = md5len; left > 55; left -= 64, WP +=16)
+			md5_block(uint, WP, digest);
+		md5_block(uint, WP, digest);
+
+		if (pdf_salt->key_length == 40) {
+			rc4_40_set_key(&rc4_ctx, key);
+			rc4(&rc4_ctx, digest, output, 16);
+			for (uint x = 0x01010101; x <= 0x13131313; x += 0x01010101) {
+				uint xor[2];
```

Comment:
Surprisingly, this array named `xor` fails build on ancient NVIDIA driver on our "bull". Renaming to `rox` makes it build and pass test.

Reply:
The name `key_xor_x` also works, so I think let's use that.

Reply:
> Surprisingly, this array named xor fails

Interesting bug. Did it fail with some descriptive error?

Reply:
Yes, IIRC it said it was expecting a `(` after `xor`, so probably had it as a function or macro.